File: internal/shadow/adapters/git_test.go
package adapters

import (
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// repoWithCommit builds a git repo with one deterministic commit and returns its HEAD SHA.
func repoWithCommit(t *testing.T, content string) string {
	t.Helper()
	dir := t.TempDir()
	require.NoError(t, exec.Command("git", "init", "-q", dir).Run())
	require.NoError(t, exec.Command("git", "-C", dir, "config", "commit.gpgsign", "false").Run())
	require.NoError(t, os.WriteFile(filepath.Join(dir, "file.txt"), []byte(content), 0o644))
	require.NoError(t, ConfigureBotIdentity(dir))
	require.NoError(t, CommitAll(dir, "shadow: fixed message"))
	sha, err := HeadSha(dir)
	require.NoError(t, err)
	return sha
}

func TestCommitDeterminismSameInputsSameSHA(t *testing.T) {
	needsTool(t, "git")
	a := repoWithCommit(t, "same")
	b := repoWithCommit(t, "same")
	assert.Regexp(t, regexp.MustCompile(`^[0-9a-f]{40}$`), a)
	assert.Equal(t, a, b)
}

func TestCommitDifferentContentDifferentSHA(t *testing.T) {
	needsTool(t, "git")
	assert.NotEqual(t, repoWithCommit(t, "one"), repoWithCommit(t, "two"))
}

File: shadow/test/git.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { execFileSync } from 'node:child_process';
import { mkdtempSync, writeFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { configureBotIdentity, commitAll, headSha } from '../src/adapters/git.ts';

async function repoWithCommit(content: string): Promise<{ dir: string; sha: string }> {
  const dir = mkdtempSync(join(tmpdir(), 'git-'));
  execFileSync('git', ['init', '-q', dir]);
  execFileSync('git', ['-C', dir, 'config', 'commit.gpgsign', 'false']);
  writeFileSync(join(dir, 'file.txt'), content);
  await configureBotIdentity(dir);
  await commitAll(dir, 'shadow: fixed message');
  return { dir, sha: await headSha(dir) };
}

describe('git adapter — reproducible commits (determinism)', () => {
  it('identical content + identity + message + dates -> identical SHA', async () => {
    const a = await repoWithCommit('same');
    const b = await repoWithCommit('same');
    try {
      assert.match(a.sha, /^[0-9a-f]{40}$/);
      assert.equal(a.sha, b.sha);
    } finally {
      rmSync(a.dir, { recursive: true, force: true });
      rmSync(b.dir, { recursive: true, force: true });
    }
  });

  it('different content -> different SHA', async () => {
    const a = await repoWithCommit('one');
    const b = await repoWithCommit('two');
    try {
      assert.notEqual(a.sha, b.sha);
    } finally {
      rmSync(a.dir, { recursive: true, force: true });
      rmSync(b.dir, { recursive: true, force: true });
    }
  });
});
